storage/src/node/mod.rs

```diff
@@ -337,59 +332,48 @@ impl Node {
                 }
             }
             Node::Leaf(l) => {
-                let first_byte: LeafFirstByte = LeafFirstByte::new(1, l.partial_path.0.len() as u8);
+                let pp_len = match l.partial_path.len() {
+                    // less than 126 nibbles
+                    len if len < LEAF_PARTIAL_PATH_LEN_OVERFLOW as usize => len as u8,
+                    _ => LEAF_PARTIAL_PATH_LEN_OVERFLOW,
+                };
+                let first_byte: LeafFirstByte = LeafFirstByte::new(1, pp_len);
 
                 const OPTIMIZE_LEAVES_FOR_SIZE: usize = 128;
                 encoded.reserve(OPTIMIZE_LEAVES_FOR_SIZE);
                 encoded.push(prefix);
                 encoded.push(first_byte.0);
 
                 // encode the partial path, including the length if it didn't fit above
-                if l.partial_path.0.len() >= 127 {
-                    encoded
-                        .write_varint(l.partial_path.len())
-                        .expect("write to array should succeed");
+                if pp_len == LEAF_PARTIAL_PATH_LEN_OVERFLOW {
+                    encoded.extend_var_int(l.partial_path.len());
```

Comment:
I wonder if the slowdown is here. `write_varint` was doing the same thing before, kind of. It allocated a stack buffer and then encoded into it, then copied the bytes, but using write_all instead of copy_from_slice:

Old:
```rust
impl<Inner: Write> VarIntWriter for Inner {
    fn write_varint<VI: VarInt>(&mut self, n: VI) -> Result<usize> {
        let mut buf = [0_u8; 10];
        let used = n.encode_var(&mut buf[..]);

        self.write_all(&buf[0..used])?;
        Ok(used)
    }
}
```
New:
```rust
    #[expect(clippy::indexing_slicing)]
    fn extend_var_int<VI: VarInt>(&mut self, int: VI) {
        let mut buf = [0u8; 10];
        let len = VarInt::encode_var(int, &mut buf);
        self.extend_from_slice(&buf[..len]);
    }
```

Reply:
The new code should be better because we're not returning an `result` which means we're also not checking the result. The `?` operator is notorious for adding fluffy code during LLVM codegen which would slow down anything not using LTO during compilation.